mantidproject/mantid
# Better way to handle ThetaIn in RROA

This PR includes the following changes to improve the way `ThetaIn` is handled in `ReflectometryReductionOneAuto`.

- Add a new property, `ThetaLogName`, to `ReflectometryReductionOneAuto`, which is used to find `ThetaIn` from the specified log value. Note that if the `ThetaIn` property is also specified, that will take precedence.
- Add another new property, `CorrectDetectors`, to make it optional whether detectors are moved to `ThetaIn` (whether specified via `ThetaIn` or `ThetaLogName`). The default is true to maintain the previous behaviour when using `ThetaIn`.
- Always pass the `theta` value through to `ReflectometryReductionOne`. If `ThetaIn` and `ThetaLogName` are not specified, `theta` is instead calculated using `SpecularReflectionCalculateTheta`. The latter does not work for some IDF setups (see linked issue) but in those cases it can now be overridden with `ThetaIn` or `ThetaLogName` without affecting the detector positions thanks to the new `CorrectDetectors` property.

The following minor usability improvements have also been made:
- Disable the `DetectorCorrectionType` drop-down if `CorrectDetectors` is false.
- Disable the `ReductionType` drop-down if `SummationType` is `SumInLambda`.
- Improve an error message to report angles in degrees rather than radians.

Fixes #20397

**To test:**

First, note that the INTER IDF has a bug which is being fixed by Max but has not been shipped yet. Make sure the linear-detector equation in `INTER_Definition.xml` is as follows for the purpose of testing this PR:
```
<logfile id="Theta" eq="3.162*tan(2*value*0.0174533)-46*0.0012" extract-single-value-as="last_value"/>
```

Now run the following script, which runs the same divergent beam reduction on an INTER linear detector with theta taken from:
- the (incorrect) internal calculation using SpecularReflectionCalculateTheta
- the (correct) value overridden by the ThetaIn property
- the (correct) value taken from the `theta` log value

```

inst='INTER'
run=43583
name=inst+str(run)
LoadISISNexus(Filename=name,OutputWorkspace=name)

instructions='45-58'
thetain=1.3
thetalog='theta'
qmin=0.008
qmax=0.5
qstep=-0.0001
theta0=thetain

# the (incorrect) internal calculation using SpecularReflectionCalculateTheta
ReflectometryReductionOneAuto(
    CorrectDetectors=0,
    InputWorkspace='INTER43583', 
    SummationType='SumInQ', ReductionType='DivergentBeam', AnalysisMode='MultiDetectorAnalysis', 
    ProcessingInstructions=instructions, MomentumTransferMin=qmin, MomentumTransferStep=qstep, MomentumTransferMax=qmax, ScaleFactor=1,
    OutputWorkspaceBinned='INTER43583_IvsQ_SRCT', 
    OutputWorkspace='INTER43583_IvsQ_unbinned', 
    OutputWorkspaceWavelength='INTER43583_IvsLam')

# the correct value passed via the ThetaIn property
ReflectometryReductionOneAuto(
    ThetaIn=thetain,
    CorrectDetectors=0,
    InputWorkspace='INTER43583', 
    SummationType='SumInQ', ReductionType='DivergentBeam', AnalysisMode='MultiDetectorAnalysis', 
    ProcessingInstructions=instructions, MomentumTransferMin=qmin, MomentumTransferStep=qstep, MomentumTransferMax=qmax, ScaleFactor=1,
    OutputWorkspaceBinned='INTER43583_IvsQ_ThetaIn_'+str(thetain), 
    OutputWorkspace='INTER43583_IvsQ_unbinned', 
    OutputWorkspaceWavelength='INTER43583_IvsLam')
    
# the correct value taken from theta in the logs
ReflectometryReductionOneAuto(
    ThetaLogName=thetalog,
    CorrectDetectors=0,
    InputWorkspace='INTER43583', 
    SummationType='SumInQ', ReductionType='DivergentBeam', AnalysisMode='MultiDetectorAnalysis', 
    ProcessingInstructions=instructions, MomentumTransferMin=qmin, MomentumTransferStep=qstep, MomentumTransferMax=qmax, ScaleFactor=1,
    OutputWorkspaceBinned='INTER43583_IvsQ_ThetaLogName_'+thetalog, 
    OutputWorkspace='INTER43583_IvsQ_unbinned', 
    OutputWorkspaceWavelength='INTER43583_IvsLam')
    
CompareWorkspaces(Workspace1='INTER43583_IvsQ_SRCT', Workspace2='INTER43583_IvsQ_ThetaIn_'+str(thetain), Tolerance=0.001)
CompareWorkspaces(Workspace1='INTER43583_IvsQ_ThetaIn_'+str(thetain), Workspace2='INTER43583_IvsQ_ThetaLogName_'+thetalog, Tolerance=0.001)

# Plot the results
lines = plot(['INTER43583_IvsQ_SRCT','INTER43583_IvsQ_ThetaIn_'+str(thetain),'INTER43583_IvsQ_ThetaLogName_'+thetalog],  0, tool='plot_spectrum')
fig = lines[0].figure()
fig.suptitle("IvsQ")
ax = fig.axes()[0]
ax.set_xscale('log')
ax.set_yscale('log')
```

The `SRCT` plot is in the wrong place because of the known issue with `SpecularReflectionCalculateTheta` for this IDF setup. The other two plots are correct:

![20397_fixed](https://user-images.githubusercontent.com/12895056/30274341-ee0829de-96f4-11e7-8d15-280e1b0227a5.png)

**Release Notes** 
*Added to the release notes*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
